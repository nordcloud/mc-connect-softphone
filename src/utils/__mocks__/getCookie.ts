export const getCookie = jest.fn().mockReturnValue('dummyCookieValue');
