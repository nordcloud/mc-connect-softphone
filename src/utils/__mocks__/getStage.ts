export const getStage = jest.fn().mockReturnValue('prod');
