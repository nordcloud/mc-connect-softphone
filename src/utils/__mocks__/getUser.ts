export const getUser = jest.fn().mockResolvedValue({
  name: 'dummyName',
  email: 'dummyEmail',
  sub: 'dummySub',
});
