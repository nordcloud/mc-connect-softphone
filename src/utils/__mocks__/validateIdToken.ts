export const validateIdToken = jest.fn().mockResolvedValue({
  active: true,
  sub: 'dummySub',
  name: 'dummyName',
  email: 'dummyEmail',
});
