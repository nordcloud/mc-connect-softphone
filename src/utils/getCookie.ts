import { APIGatewayProxyEventV2 } from 'aws-lambda';
import cookie from 'cookie';

export function getCookie(event: APIGatewayProxyEventV2, cookieName: string) {
  if (!event.cookies || !event.cookies.length) {
    return undefined;
  }

  const cookies = cookie.parse(event.cookies.join(';'));

  return cookies[cookieName] as string | undefined;
}
