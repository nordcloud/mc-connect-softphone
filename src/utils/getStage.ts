import { Stage } from '../types';
import { assertStageName } from './assertStageName';

export function getStage(): Stage {
  const { STAGE } = process.env;

  assertStageName(STAGE);

  return STAGE;
}
