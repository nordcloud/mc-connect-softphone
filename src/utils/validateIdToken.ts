import fetch from 'node-fetch';

import { OauthIntrospectResponse } from '../types';
import { getStageConsts } from './getStageConsts';

const { OAUTH_BASE_URL, OAUTH_CLIENT_ID } = getStageConsts();

export async function validateIdToken(idToken: string) {
  console.log('Validating ID token');

  const res = await fetch(`${OAUTH_BASE_URL}/introspect`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body: new URLSearchParams({
      client_id: OAUTH_CLIENT_ID,
      token: idToken,
      token_type_hint: 'id_token',
    }),
  });

  const data = (await res.json()) as OauthIntrospectResponse;

  if ('error' in data) {
    throw new Error(`${data.error}: ${data.error_description}`);
  }

  return data;
}
