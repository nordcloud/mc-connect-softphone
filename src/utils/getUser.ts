import { APIGatewayProxyEventV2 } from 'aws-lambda';

import { ID_TOKEN_COOKIE } from '../consts';
import { userMock } from '../mockData';
import { getCookie } from './getCookie';
import { getStage } from './getStage';
import { validateIdToken } from './validateIdToken';

export async function getUser(event: APIGatewayProxyEventV2) {
  const isMockMode =
    getStage() === 'local' &&
    Object.keys(Object(event.queryStringParameters)).includes('mock_login');

  if (isMockMode) {
    return userMock;
  }

  const idToken = getCookie(event, ID_TOKEN_COOKIE);

  if (!idToken) {
    return undefined;
  }

  const validationResult = await validateIdToken(idToken);

  if (!validationResult.active) {
    return undefined;
  }

  const { sub, name, email } = validationResult;

  return {
    sub,
    name,
    email,
  };
}
