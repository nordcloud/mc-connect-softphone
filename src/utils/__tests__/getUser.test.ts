import { APIGatewayProxyEventV2 } from 'aws-lambda';

import { getCookie } from '../getCookie';
import { getStage } from '../getStage';
import { getUser } from '../getUser';
import { validateIdToken } from '../validateIdToken';

jest.mock('../getCookie');
jest.mock('../getStage');
jest.mock('../validateIdToken');

const event = {} as APIGatewayProxyEventV2;

describe('getuser', () => {
  it('calls getCookie with correct args', async () => {
    await getUser(event);
    expect(jest.mocked(getCookie).mock.calls).toMatchSnapshot();
  });

  it('calls validateIdToken with correct args', async () => {
    await getUser(event);
    expect(jest.mocked(validateIdToken).mock.calls).toMatchSnapshot();
  });

  it('resolves with a correct value', () => {
    expect(getUser(event)).resolves.toMatchSnapshot();
  });

  it('resolves with a correct value when ID token is missing', () => {
    jest.mocked(getCookie).mockReturnValueOnce(undefined);
    expect(getUser(event)).resolves.toMatchSnapshot();
  });

  it('resolves with a correct value when ID token is invalid', () => {
    jest.mocked(validateIdToken).mockResolvedValueOnce({ active: false });
    expect(getUser(event)).resolves.toMatchSnapshot();
  });

  it('resolves with a correct value when running in "local" stage', () => {
    jest.mocked(getStage).mockReturnValueOnce('local');
    expect(getUser(event)).resolves.toMatchSnapshot();
  });
});
