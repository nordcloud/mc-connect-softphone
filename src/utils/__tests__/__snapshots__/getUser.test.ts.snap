// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getuser calls getCookie with correct args 1`] = `
[
  [
    {},
    "idToken",
  ],
]
`;

exports[`getuser calls validateIdToken with correct args 1`] = `
[
  [
    "dummyCookieValue",
  ],
]
`;

exports[`getuser resolves with a correct value 1`] = `
{
  "email": "dummyEmail",
  "name": "dummyName",
  "sub": "dummySub",
}
`;

exports[`getuser resolves with a correct value when ID token is invalid 1`] = `undefined`;

exports[`getuser resolves with a correct value when ID token is missing 1`] = `undefined`;

exports[`getuser when running in mock mode resolves with a correct value in "local" stage 1`] = `
{
  "email": "mocked.user@domain.com",
  "idToken": "mockedUserIdToken",
  "name": "Mocked User",
  "sub": "mockedUserSub",
}
`;

exports[`getuser when running in mock mode resolves with a correct value in "prod" stage 1`] = `
{
  "email": "dummyEmail",
  "name": "dummyName",
  "sub": "dummySub",
}
`;
